src/Microsoft.AspNetCore.Sockets.Client/Connection.cs

```diff
@@ -7,99 +7,136 @@
         private readonly ILogger _logger;
+        private readonly ILoggerFactory _loggerFactory;
 
         public Uri Url { get; }
 
-        // TODO: Review. This is really only designed to be used from ConnectAsync
-        private Connection(Uri url, ITransport transport, IChannelConnection<Message> transportChannel, ILogger logger)
+        public Connection(Uri url)
+            : this(url, null)
+        { }
+
+        public Connection(Uri url, ILoggerFactory loggerFactory)
         {
-            Url = url;
+            if (url == null)
+            {
+                throw new ArgumentNullException(nameof(url));
+            }
 
-            _logger = logger;
-            _transport = transport;
-            _transportChannel = transportChannel;
+            Url = url;
+            _loggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
+            _logger = _loggerFactory.CreateLogger<Connection>();
         }
 
         public ReadableChannel<Message> Input => _transportChannel.Input;
         public WritableChannel<Message> Output => _transportChannel.Output;
 
-        public void Dispose()
+        public Task StartAsync()
         {
-            _transport.Dispose();
+            return StartAsync(null, null);
         }
 
-        public static Task<Connection> ConnectAsync(Uri url, ITransport transport) => ConnectAsync(url, transport, new HttpClient(), NullLoggerFactory.Instance);
-        public static Task<Connection> ConnectAsync(Uri url, ITransport transport, ILoggerFactory loggerFactory) => ConnectAsync(url, transport, new HttpClient(), loggerFactory);
-        public static Task<Connection> ConnectAsync(Uri url, ITransport transport, HttpClient httpClient) => ConnectAsync(url, transport, httpClient, NullLoggerFactory.Instance);
+        public Task StartAsync(ITransport transport)
+        {
+            return StartAsync(transport, null);
+        }
 
-        public static async Task<Connection> ConnectAsync(Uri url, ITransport transport, HttpClient httpClient, ILoggerFactory loggerFactory)
+        public async Task StartAsync(ITransport transport, HttpClient httpClient)
         {
-            if (url == null)
+            if (Interlocked.CompareExchange(ref _connectionState, ConnectionState.Connecting, ConnectionState.Disconnected)
+                != ConnectionState.Disconnected)
             {
-                throw new ArgumentNullException(nameof(url));
+                throw new InvalidOperationException("Cannot start an already running connection.");
             }
 
-            if (transport == null)
+            if (httpClient == null)
             {
-                throw new ArgumentNullException(nameof(transport));
+                // TODO: httpClient needs to be disposed properly (after long polling is not the default transport)
+                httpClient = new HttpClient();
             }
 
-            if (httpClient == null)
+            var connectionId = await Negotiate(httpClient);
```

Comment:
We're going to have to go through and add `ConfigureAwait(false)` on a bunch of these calls